Multiaddr duplicated p2p protocol in relay reservation (#2696)

stormshield-pj50 opened the pull request:
# Description
In protocols/relay/src/v2/relay.rs do not append P2p protocol if already present.

Now the built multiaddr does not contain a dupped P2p protocol.  Logs are OK:
```
[2022-06-07T08:38:16Z INFO  client] Relay accepted our reservation request.
[2022-06-07T08:38:16Z INFO  client] Listening on "/ip4/$RELAY_SERVER_IP/tcp/4001/p2p/12D3KooWDpJ7As7BWAwRMfu1VU2WCqNjvq387JEYKDBj4kx6nXTN/p2p-circuit/p2p/12D3KooWPjceQrSwdWXPyLLeABRXmuqt69Rg3sBYbU1Nft9HyQ6X"
```

## Open Questions

## Change checklist

- [ X] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

stormshield-pj50 committed:
Multiaddr duplicated p2p protocol in relay reservation (#2696)

mxinden commented on protocols/relay/src/v2/relay.rs:
```diff
@@ -757,8 +757,12 @@ impl Action {
+                            let p2p_proto =
+                                Protocol::P2p(*poll_parameters.local_peer_id().as_ref());
+                            match a.addr.iter().last() {
+                                Some(p) if p == p2p_proto => a.addr,
```
I don't think we need to check whether the `PeerId` is our own. This check should happen earlier in the process, in case it is not already happening.

mxinden reviewed (commented):
Thanks for tackling this.

Small suggestions. What do you think of the diff below?

``` diff
diff --git a/protocols/relay/src/v2/relay.rs b/protocols/relay/src/v2/relay.rs
index 8d8bb75c..ed5fe6ca 100644
--- a/protocols/relay/src/v2/relay.rs
+++ b/protocols/relay/src/v2/relay.rs
@@ -756,9 +756,13 @@ impl Action {
                     inbound_reservation_req,
                     addrs: poll_parameters
                         .external_addresses()
-                        .map(|a| {
-                            a.addr
-                                .with(Protocol::P2p((*poll_parameters.local_peer_id()).into()))
+                        .map(|a| a.addr)
+                        // Add local peer ID in case it isn't present yet.
+                        .filter_map(|a| match a.iter().last()? {
+                            Protocol::P2p(_) => Some(a),
+                            _ => Some(
+                                a.with(Protocol::P2p(*poll_parameters.local_peer_id().as_ref())),
+                            ),
                         })
                         .collect(),
                 }),
```

Also, would you mind including a changelog entry and a `libp2p-relay` patch version bump?

stormshield-pj50 committed:
Multiaddr duplicated p2p protocol in relay reservation (#2696)
- Minor fixes after review
- Bump libp2p-relay Cargo.toml version to 0.9.2
- Add entry in Changelog for upcoming 0.9.2 version

mxinden committed:
Merge branch 'master' into issue_2696_multiaddr_dup_p2p_protocol

mxinden reviewed (approved):
Sorry for the delay here. Thanks @pj50. 

I am planning to cut a new release soonish. In case you need this earlier, let me know.
